File: raspberry/SmartHome/thirdpartydevices/src/main/java/smarthome/raspberry/thirdpartydevices/xiaomi/gateway/device/DoorWindowSensor.java
package smarthome.raspberry.thirdpartydevices.xiaomi.gateway.device;

import org.json.JSONException;
import org.json.JSONObject;

import static smarthome.library.common.constants.DeviceTypes.DOOR_WINDOW_SENSOR_TYPE;

public class DoorWindowSensor extends ExtendedDevice  {

    private OnStateChangeListener listener;

    public DoorWindowSensor(String sid) {
        super(sid, DOOR_WINDOW_SENSOR_TYPE);
    }

    public DoorWindowSensor(String sid, OnStateChangeListener listener) {
        this(sid);
        setListener(listener);
    }


    public void setListener(OnStateChangeListener listener) {
        this.listener = listener;
    }


    @Override
    public void parseData(String cmd) {
        try {
            JSONObject o = new JSONObject(cmd);

            if(!o.isNull(STATUS_KEY)) {
                setStatus(o.getString(STATUS_KEY));

                report(getStatus());
            }

            if(!o.isNull(VOLTAGE_KEY)) {
                String vol = o.getString(VOLTAGE_KEY);
                setVoltage(Float.parseFloat(vol) / 1000);
            }

        } catch (JSONException | NumberFormatException e) {
            e.printStackTrace();
        }
    }


    private void report(String state) {
        if(listener != null)
            listener.onStateChanged(state);
    }

    public interface OnStateChangeListener {

        void onStateChanged(String state);
    }
}

File: raspberry/SmartHome/thirdpartydevices/src/main/java/smarthome/raspberry/thirdpartydevices/xiaomi/gateway/device/DoorWindowSensor.kt
package smarthome.raspberry.thirdpartydevices.xiaomi.gateway.device

import org.json.JSONException
import org.json.JSONObject
import smarthome.library.common.constants.DeviceTypes.DOOR_WINDOW_SENSOR_TYPE
import smarthome.library.common.constants.GATEWAY_DOOR_WINDOW_CONTROLLER
import smarthome.library.common.constants.STATUS_KEY
import smarthome.raspberry.thirdpartydevices.xiaomi.gateway.controller.DoorWindowController
import smarthome.raspberry.thirdpartydevices.xiaomi.gateway.controller.VoltageController

class DoorWindowSensor(sid: String, gatewaySid: String) : GatewayDevice(sid, DOOR_WINDOW_SENSOR_TYPE, gatewaySid) {

    init {
        addControllers(DoorWindowController(this),
                VoltageController(this))
    }

    override fun parseData(json: String) {
        try {
            val o = JSONObject(json)

            if (!o.isNull(STATUS_KEY))
                getControllerByType(GATEWAY_DOOR_WINDOW_CONTROLLER).state = o.getString(STATUS_KEY)

            setVoltage(o)

            super.parseData(json)
        } catch (e: JSONException) {
            reportDataParseError(e)
        }
    }
}